llama-index-core/tests/workflow/test_service.py

```diff
@@ -0,0 +1,52 @@
+import pytest
+
+from llama_index.core.workflow.decorators import step
+from llama_index.core.workflow.events import Event, StartEvent, StopEvent
+from llama_index.core.workflow.workflow import Workflow
+from llama_index.core.workflow.service import ServiceManager
+
+
+class ServiceWorkflow(Workflow):
+    """This wokflow is only responsible to generate a number, it knows nothing about the caller."""
+
+    @step()
+    async def generate(self, ev: StartEvent) -> StopEvent:
+        return StopEvent(result=42)
+
+
+class NumGenerated(Event):
+    """To be used in the dummy workflow below."""
+
+    num: int
+
+
+class DummyWorkflow(Workflow):
+    """
+    This workflow needs a number, and it calls another workflow to get one.
+    A service named "service_workflow" must be added to `DummyWorkflow` for
+    the step to be able to use it (see below).
+    This step knows nothing about the other workflow, it gets an instance
+    and it only knows it has to call `run` on that instance.
+    """
+
+    @step(services=["service_workflow"])
+    async def get_a_number(self, svc: Workflow, ev: StartEvent) -> NumGenerated:
+        res = await svc.run()
+        return NumGenerated(num=int(res))
+
+    @step()
+    async def multiply(self, ev: NumGenerated) -> StopEvent:
+        return StopEvent(ev.num * 2)
+
+
+@pytest.mark.asyncio()
+async def test_e2e():
+    # We are responsible for passing the ServiceWorkflow instances to the dummy workflow
+    # and give it a name, in this case "service_workflow"
+    wf = DummyWorkflow(
+        service_manager=ServiceManager.from_workflows(
```

Comment:
If it helps, keep in mind this would only be needed in case of nested workflows :) 

What I personally don't like:
- I don't mind the extra import but it's one more concept to learn (even if users will thank me later when debugging llama-agents eheh)
What I like:
- you can just pass `service_name=ServiceInstance()` that wouldn't be possible in the constructor. 
 
The alternative would be passing in a dictionary like `wf = DummyWorkflow(services={"service_workflow": ServiceWorkflow()})` or introducing a "service descriptor" object that would cause similar UX problems.

Reply:
I removed `ServiceManager` from the public API and brought back the `add_services` method